src/string.cr

```diff
@@ -891,11 +891,30 @@ class String
     end
   end
 
+  # Returns the `Char?` at the given *index*.
+  #
+  # Negative indices can be used to start counting from the end of the string.
+  #
+  # See '#[]'
+  #
+  # ```
+  # "hello"[0]  # => 'h'
+  # "hello"[1]  # => 'e'
+  # "hello"[-1] # => 'o'
+  # "hello"[-2] # => 'l'
+  # "hello"[5]  # => nil
+  # ```
   def []?(index : Int) : Char?
     char_at(index) { nil }
   end
 
-  def []?(str : String | Char)
+  # Returns `String?` of matching *given* if the given *str* is in *self*.
+  #
+  # ```
+  # "crystal"["cry"]?  # => "cry"
+  # "crystal"["ruby"]? # => nil
+  # ```
+  def []?(str : String | Char) : String?
```

Comment:
The return type restriction is incorrect. If `str` is a `Char`, the return type is `Char?`.
```suggestion
  def []?(str : String | Char)
```
This should probably just be two separate methods. Or entirely remove? I don't know about practical use cases.